Sources/NIO/EventLoop.swift

```diff
@@ -50,6 +50,67 @@ public struct Scheduled<T> {
     }
 }
 
+/// Returned once a task was scheduled to be repeatedly executed on the `EventLoop`.
+///
+/// A `RepeatedTask` allows the user to `cancel()` the repeated scheduling of further tasks.
+public final class RepeatedTask {
+    private let delay: TimeAmount
+    private var scheduled: Scheduled<EventLoopFuture<Void>>?
+    private var task: (() throws -> EventLoopFuture<Void>)?
+    private let lock = Lock()
```

Comment:
I don't think we want to use a lock here: instead, we should use the event loop that this `RepeatedTask` belongs to to manage synchronization.

The TL;DR for this is that locking is hard, and right now this isn't handling it very well. In particular, `begin` does not handle the synchronisation correctly. The other advantage is that with using the parent event loop we can easily check whether we are already holding the "lock" and so provide a fast-path that does not involve cross-loop dispatch.

Reply:
I was thinking about the `begin` method as well but since this is only used internally and the user can only call cancel after `reschedule` was triggered once it should be fine. 

Would synchronization over the EL work with `executing` begin, reschedule and cancel on the EL? @Lukasa

Reply:
Yup, should work fine. In fact, only `cancel` should require the explicit synchronisation: both `begin` and `cancel` should by definition always be called on the loop, and so can just degenerately `assert` that that's what's actually happening.